src/main/python/deep_qa/layers/noisy_or.py

```diff
@@ -0,0 +1,93 @@
+from keras import backend as K
+from keras.layers import Layer
+from keras.constraints import Constraint
+from keras.regularizers import L1L2Regularizer
+from overrides import overrides
+
+class BetweenZeroAndOne(Constraint):
+    """
+    Constrains the weights to be between zero and one
+    """
+    def __call__(self, p):
+        # Clip values less than or equal to zero to epsilon
+        p *= K.cast(p >= 0., K.floatx())
+        leaky_zeros_mask = K.epsilon() * K.cast(K.equal(p, K.zeros_like(p)), 'float32')
+        p = p + leaky_zeros_mask
+        # Clip values greater to 1 to 1
+        p *= K.cast(p <= 1., K.floatx())
+        leaky_ones_mask = K.cast(K.equal(p, K.zeros_like(p)), K.floatx())
+        p += leaky_ones_mask
+        return p
+
+
+class NoisyOr(Layer):
+    r"""
+    This layer takes as input a tensor of probabilities and calculates the noisy-or probability
+    across a given axis based on the noisy-or equation:
+        :math:`p(x) = 1 - \prod_{i=1:N}(1 - q * p(x|y_n))`
+    where q is the noise parameter
+
+    Inputs:
+        - probabilities, shape ``(batch, ..., N, ...)``,
+          where N is the number of y's in the above equation
+          (i.e. the number of probabilities being combined in the product),
+          in the dimension corresponding to the specified axis
```

Comment:
Period at the end here.  It might look ok to read now, but if someone adds another sentence in a line below this, it'll get really hard to read.  Also, could you add a sentence about masking?  I would make it something like:

```
- probabilities, shape ``(batch, ..., N, ...)``, with optional mask of the same shape, where N...
```

And similarly in the outputs:

```
- X, shape ``(batch, ..., ...)``, with an optional mask of the same shape, one less ... [then after the text you have:] The output mask is the result of `K.any()` on the input mask, along the specified axis.
```

The wording of that last one got a little awkward with all of the clauses in that sentence, but I think you get the idea.